builq_test.go

```diff
@@ -2,46 +2,90 @@ package builq
 
 import (
 	"errors"
+	"strings"
 	"testing"
 )
 
-func TestBuilder(t *testing.T) {
-	t.Run("unsupported verb", func(t *testing.T) {
-		var b Builder
-		b.Addf("SELECT * FROM %v", "users").
-			Addf("LIMIT 100;")
+func TestColumns(t *testing.T) {
+	cols := Columns{"foo", "bar", "baz"}
 
-		if _, _, err := b.Build(); err == nil {
-			t.Errorf("must be error")
-		} else if want := "unsupported verb v"; err.Error() != want {
-			t.Errorf("\nhave: %v\nwant: %v", err, want)
+	t.Run("string", func(t *testing.T) {
+		const want = "foo, bar, baz"
+		if got := cols.String(); got != want {
+			t.Errorf("got %q; want %q", got, want)
 		}
 	})
 
-	t.Run("different placeholders", func(t *testing.T) {
-		var b Builder
-		b.Addf("WHERE foo = %$ AND bar = %?", 1, 2)
-
-		if _, _, err := b.Build(); !errors.Is(err, errMixedPlaceholders) {
-			t.Errorf("\nhave: %v\nwant: %v", err, errMixedPlaceholders)
+	t.Run("prefixed", func(t *testing.T) {
+		const want = "x.foo, x.bar, x.baz"
+		if got := cols.Prefixed("x."); got != want {
+			t.Errorf("got %q; want %q", got, want)
 		}
 	})
+}
 
-	t.Run("different placeholders in slices", func(t *testing.T) {
-		var b Builder
-		b.Addf("WHERE foo = %+$ AND bar = %+?", 1, 2)
+func TestBuilder(t *testing.T) {
+	test := func(name string, wantErr error, format string, args ...any) {
+		t.Helper()
+		t.Run(name, func(t *testing.T) {
+			t.Helper()
+			var b Builder
+			b.Addf(format, args...)
+			_, _, err := b.Build()
+			if !errors.Is(err, wantErr) {
+				t.Errorf("\nhave: %v\nwant: %v", err, wantErr)
+			}
+		})
+	}
+
+	test("unsupported verb", errUnsupportedVerb, "SELECT * FROM %v", "users")
+	test("mixed placeholders", errMixedPlaceholders, "WHERE foo = %$ AND bar = %?", 1, 2)
+	test("non-slice argument", errNonSliceArgument, "WHERE foo = %+$", 1)
+	test("non-slice argument (batch)", errNonSliceArgument, "WHERE foo = %#$", 1)
+}
 
-		if _, _, err := b.Build(); !errors.Is(err, errNonSliceArgument) {
-			t.Errorf("\nhave: %v\nwant: %v", err, errNonSliceArgument)
+func FuzzBuilder(f *testing.F) {
+	f.Add("SELECT %s FROM %s", "*", "users")
+	f.Add("SELECT * FROM %s WHERE name = %$", "users", "john")
+	f.Add("SELECT * FROM users WHERE name = %$ AND surname = %$", "john", "doe")
+
+	// queries that won't return an error (probably should):
+	// f.Add("%$%%$%*", "0", "0")
```

Comment:
Currently `Builder` won't report a lot of invalid cases like `%[` `%]`, because the the `Format` method won't be called for them. Left it as is for now, please run `-fuzz` for more examples